README.md
# **rn-popup-modal**

Display visual popup modal

## Features

- Support in React Native for both Android and iOS
- Possible to wrap more than one child component
- Easy to use
- Some customizations

## Demo

<img src=".github/demo.gif" height="500" />

## Installation

- Using **Npm**

```bash
npm install rn-popup-modal --save
```

- Using **Yarn**

```bash
yarn add rn-popup-modal
```

## Example

```jsx
import RNPopupModal from 'rn-popup-modal';

<RNPopupModal ref={ref => (this._modal = ref)} title='The Pyrenees'>
  <Image
    source={{
      uri:
        'https://www.smartertravel.com/uploads/2012/03/stm4f4fa817459b420120301.jpg'
    }}
    style={styles.image}
    resizeMode='cover'
  />
  <Text style={styles.title}>
    Best Hidden Destination for Mountain Lovers: The Pyrenees
  </Text>

  <Text style={{ fontSize: 16 }}>
    “The Pyrenees in southern France and northern Spain [are] everything people
    crave about the Alps with a fraction of the people. [They have so much to
    offer, including] incredible high-country national parks in Gavarnie,
    Ordesa, and Aiguestortes; hearty regional cuisine; the running of the bulls
    in Pamplona—and even duty-free shopping in tiny Andorra.”
  </Text>
</RNPopupModal>

// Call the method to show the popup modal
_modal.show();
```

## Usage

### Available props

| Name                            | Type   | Default         | Description                                                        |
| ------------------------------- | ------ | --------------- | ------------------------------------------------------------------ |
| title                           | string |                 | String as title displayed on top header of the modal               |
| titleStyle                      | style  |                 | Style of the title Text component                                  |
| blockBackgroundColor            | color  | rgba(0,0,0,0.5) | Background color of the blocking component behind the popup window |
| backgroundColor                 | color  | #ffffff         | Background color of the popup modal window                         |
| windowContainerStyle            | style  |                 | Container style for the popup modal window                         |
| headerLineStyle                 | style  |                 | Style of divider line between header and body                      |
| headerContainerStyle            | style  |                 | Container style for the header                                     |
| scrollViewStyle                 | style  |                 | Scroll view style                                                  |
| scrollViewContentContainerStyle | style  |                 | Content container style for the scroll view                        |
| transitionDuration              | number | 200             | Transition duration to animate the popup modal                     |
| transitionFromScale             | number | 1.1             | Start scale for transiting the popup modal                         |

also inherit [Modal Props](https://facebook.github.io/react-native/docs/modal)

### Available instance method

| Name | Description          |
| ---- | -------------------- |
| show | Show the popup modal |

## License

- [MIT](LICENSE)


CHANGELOG.md
# 1.0.0 (2019-08-27)

Initial version

examples/Example.js
import React, { Fragment } from 'react';
import { SafeAreaView, StyleSheet, Button, Text, Image } from 'react-native';
import RNPopupModal from '../src/RNPopupModal';

export default class Example extends React.Component {
  render() {
    return (
      <Fragment>
        <SafeAreaView style={styles.container}>
          <Button title="The Pyrenees" onPress={this._onPressed} />
        </SafeAreaView>
        <RNPopupModal ref={ref => (this._modal = ref)} title="The Pyrenees">
          <Image
            source={{
              uri:
                'https://www.smartertravel.com/uploads/2012/03/stm4f4fa817459b420120301.jpg'
            }}
            style={styles.image}
            resizeMode="cover"
          />
          <Text style={styles.title}>
            Best Hidden Destination for Mountain Lovers: The Pyrenees
          </Text>

          <Text style={{ fontSize: 16 }}>
            “The Pyrenees in southern France and northern Spain [are] everything
            people crave about the Alps with a fraction of the people. [They
            have so much to offer, including] incredible high-country national
            parks in Gavarnie, Ordesa, and Aiguestortes; hearty regional
            cuisine; the running of the bulls in Pamplona—and even duty-free
            shopping in tiny Andorra.”
          </Text>
        </RNPopupModal>
      </Fragment>
    );
  }

  _onPressed = () => {
    this._modal && this._modal.show();
  };
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#fff',
    padding: 20,
    justifyContent: 'center',
    alignItems: 'center'
  },
  image: {
    width: '100%',
    height: 200,
    backgroundColor: 'grey',
    marginBottom: 10
  },
  title: {
    color: 'orange',
    marginVertical: 10,
    fontSize: 20
  }
});


src/RNPopupModal.js
import React from 'react';
import {
  View,
  Text,
  Modal,
  SafeAreaView,
  ScrollView,
  TouchableWithoutFeedback,
  StyleSheet,
  Dimensions
} from 'react-native';
import { RNTweenView, TweenEasing } from 'rn-tween';
import {
  widthPercentageToDP as wp,
  heightPercentageToDP as hp
} from 'react-native-responsive-screen';
import PropTypes from 'prop-types';

const defaultRadius = Math.min(wp(2), hp(2));
const hitSlopOffset = Math.min(wp(2), hp(2));
const defaultDuration = 200;
const defaultFromScale = 1.1;

const CloseButton = ({ onPress }) => {
  const size = Math.min(wp(6), hp(6));
  const style = {
    position: 'absolute',
    width: size,
    height: size,
    end: -size / 2.0,
    top: -size / 2.0,
    borderRadius: size / 2.0,
    backgroundColor: '#000',
    alignItems: 'center',
    justifyContent: 'center'
  };

  return (
    <TouchableWithoutFeedback
      hitSlop={{
        top: hitSlopOffset,
        bottom: hitSlopOffset,
        left: hitSlopOffset,
        right: hitSlopOffset
      }}
      onPress={onPress}>
      <View style={style}>
        <Text
          style={{
            color: '#fff'
          }}>
          x
        </Text>
      </View>
    </TouchableWithoutFeedback>
  );
};
const HeaderLine = ({ style }) => (
  <View style={[defaultStyles.headerLine, StyleSheet.flatten(style)]} />
);

export default class RNPopupModal extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      visible: props.visibleOnMount ? true : false
    };

    const dimensions = Dimensions.get('screen');
    this.state.width = dimensions.width;
    this.state.height = dimensions.height;
  }

  componentDidMount() {
    Dimensions.addEventListener('change', this._onDimensionChange);
  }

  _onDimensionChange = ({ screen }) => {
    if (this._willUnmount) {
      return;
    }

    this.setState({
      width: screen.width,
      height: screen.height
    });
  };

  componentWillUnmount() {
    this._willUnmount = true;
    Dimensions.removeEventListener('change', this._onDimensionChange);
  }

  show() {
    this.setState({ visible: true });
  }

  render() {
    const {
      blockBackgroundColor,
      windowContainerStyle,
      visible,
      onRequestClose,
      transitionDuration,
      transitionFromScale,
      backgroundColor,
      ...restProps
    } = this.props;
    const { width, height } = this.state;
    const duration =
      transitionDuration === null || transitionDuration === undefined
        ? defaultDuration
        : transitionDuration;
    const fromScale =
      transitionFromScale === null || transitionFromScale === undefined
        ? defaultFromScale
        : transitionFromScale;

    return (
      <Modal
        {...restProps}
        visible={this.state.visible}
        onRequestClose={this._onRequestClose}>
        <SafeAreaView
          style={{
            flex: 1,
            justifyContent: 'center',
            alignItems: 'center',
            backgroundColor: blockBackgroundColor
              ? blockBackgroundColor
              : 'transparent'
          }}>
          <TouchableWithoutFeedback onPress={this._onRequestClose}>
            <View style={StyleSheet.absoluteFillObject} />
          </TouchableWithoutFeedback>

          <RNTweenView
            autoStartName="transit"
            initialConfig={{
              transit: {
                mode: 'parallel',
                configs: [
                  {
                    from: 0,
                    to: 1,
                    duration,
                    styleProperty: 'opacity'
                  },
                  {
                    from: fromScale,
                    to: 1,
                    duration,
                    styleProperty: 'scale',
                    easing: TweenEasing.linear
                  }
                ]
              }
            }}
            style={{
              ...StyleSheet.flatten(windowContainerStyle),
              width: width * 0.8,
              height: height * 0.8
            }}>
            <View style={{ ...defaultStyles.visualWindow, backgroundColor }}>
              {this._renderHead()}
              {this._renderBody()}
            </View>

            <CloseButton onPress={this._onRequestClose} />
          </RNTweenView>
        </SafeAreaView>
      </Modal>
    );
  }

  _renderHead() {
    const {
      headerContainerStyle,
      titleStyle,
      title,
      headerLineStyle
    } = this.props;

    return [
      <View
        key={1}
        style={[
          defaultStyles.headerContainer,
          StyleSheet.flatten(headerContainerStyle)
        ]}>
        <Text
          style={[defaultStyles.title, StyleSheet.flatten(titleStyle)]}
          numberOfLines={1}>
          {title}
        </Text>
      </View>,
      <HeaderLine key={2} style={headerLineStyle} />
    ];
  }

  _renderBody() {
    const {
      scrollViewStyle,
      children,
      scrollViewContentContainerStyle
    } = this.props;

    return (
      <ScrollView
        style={[defaultStyles.scrollView, StyleSheet.flatten(scrollViewStyle)]}
        contentContainerStyle={[
          defaultStyles.scrollViewContentContainer,
          StyleSheet.flatten(scrollViewContentContainerStyle)
        ]}>
        {children}
      </ScrollView>
    );
  }

  _onRequestClose = () => {
    this.setState({ visible: false });
  };
}

RNPopupModal.defaultProps = {
  blockBackgroundColor: 'rgba(0,0,0,0.5)',
  supportedOrientations: [
    'landscape',
    'portrait',
    'landscape-left',
    'landscape-right',
    'portrait-upside-down'
  ],
  transparent: true,
  backgroundColor: '#fff'
};

RNPopupModal.propTypes = {
  title: PropTypes.string,
  blockBackgroundColor: PropTypes.string,
  titleStyle: PropTypes.oneOfType([PropTypes.object, PropTypes.array]),
  windowContainerStyle: PropTypes.oneOfType([
    PropTypes.object,
    PropTypes.array
  ]),
  headerLineStyle: PropTypes.oneOfType([PropTypes.object, PropTypes.array]),
  headerContainerStyle: PropTypes.oneOfType([
    PropTypes.object,
    PropTypes.array
  ]),
  scrollViewStyle: PropTypes.oneOfType([PropTypes.object, PropTypes.array]),
  scrollViewContentContainerStyle: PropTypes.oneOfType([
    PropTypes.object,
    PropTypes.array
  ]),
  transitionDuration: PropTypes.number,
  transitionFromScale: PropTypes.number,
  backgroundColor: PropTypes.string
};

const defaultStyles = {
  visualWindow: {
    flex: 1,
    backgroundColor: '#fff',
    borderRadius: defaultRadius,
    overflow: 'hidden'
  },
  headerContainer: {
    paddingHorizontal: Math.min(wp(2), hp(2)),
    paddingVertical: Math.min(wp(2), hp(2)),
    alignItems: 'center',
    justifyContent: 'center'
  },
  title: {
    fontSize: Math.min(wp(5), hp(5)),
    fontWeight: 'bold',
    textAlignVertical: 'center'
  },
  scrollView: {
    flex: 1
  },
  scrollViewContentContainer: {
    padding: Math.min(wp(2), hp(2))
  },
  headerLine: {
    width: '100%',
    height: StyleSheet.hairlineWidth,
    backgroundColor: 'grey'
  }
};
